File: Creational Pattern/Factory Method/cpp/factory_method.cpp
#include <string>
#include <iostream>

class Product
{
public:
    virtual ~Product() {}
    virtual std::string Operation() const = 0;
};

class ConcreteProduct1 : public Product
{
public:
    std::string Operation() const override
    {
        return "{Result of the ConcreteProduct1}";
    }
};

class ConcreteProduct2 : public Product
{
public:
    std::string Operation() const override
    {
        return "{Result of the ConcreteProduct2}";
    }
};

class Creater
{
public:
    virtual ~Creater(){};
    virtual Product *FactoryMethod() const = 0;

    std::string SomeOperation() const
    {
        Product *product = this->FactoryMethod();
        std::string result = "Creator: The same creator's code has just worked with " + product->Operation();

        delete product;
        return result;
    }
};

class ConcreteCreator1 : public Creater
{
public:
    Product *FactoryMethod() const override
    {
        return new ConcreteProduct1();
    }
};

class ConcreteCreator2 : public Creater
{
public:
    Product *FactoryMethod() const override
    {
        return new ConcreteProduct2();
    }
};

void ClientCode(const Creater &creater)
{
    std::cout << "Client: I'm not aware of the creator's class, but it still works.\n"
              << creater.SomeOperation() << std::endl;
}

int main()
{
    std::cout << "App: Launched with the ConcreateCreater1.\n";
    Creater *creater = new ConcreteCreator1();
    ClientCode(*creater);
    std::cout << std::endl;
    std::cout << "App: Launched with the ConcreateCreater2.\n";
    Creater *creater2 = new ConcreteCreator2();
    ClientCode(*creater2);

    delete creater;
    delete creater2;

    return 0;
}

File: Creational Pattern/Factory Method/python/factory_method.py
from abc import ABC, abstractmethod


class Creator(ABC):
    @abstractmethod
    def factory_method(self):
        pass

    def some_operation(self) -> str:

        product = self.factory_method()

        result = f"Creator: The same creator's code has just worked with {product.operation()}"

        return result


class Product(ABC):
    @abstractmethod
    def operation(self) -> str:
        pass


class ConcreteProduct1(Product):
    def operation(self) -> str:
        return "{Result of the ConcreteProduct1}"


class ConcreteProduct2(Product):
    def operation(self) -> str:
        return "{Result of the ConcreteProduct2}"


class concreteCreator1(Creator):
    def factory_method(self) -> Product:
        return ConcreteProduct1()


class concreteCreator2(Creator):
    def factory_method(self) -> Product:
        return ConcreteProduct2()


def client_code(creator: Creator) -> None:

    print(f"Client: I'm not aware of the creator's class, but it still works.\n"
          f"{creator.some_operation()}", end="")


if __name__ == "__main__":
    print("App: Launched with the ConcreteCreator1.")
    client_code(concreteCreator1())
    print("\n")

    print("App: Launched with the ConcreteCreator1.")
    client_code(concreteCreator2())
    print("\n")
